### ISSUE TITLE
Access to Homeworks is TCP/IP (no serial port)?

### ISSUE THREAD

**Issue by mikmort** [dubnom/pyhomeworks#6]
I have both a regular IP connection to my Homeworks 8 system an a straight serial connection. However it looks like this solution only support ethernet over serial?

Is there a way in the configuration to get it the straight telnet IP address? Something:

192.168.1.23
Port: 23

?

**Comment by dubnom** [comment]
I didn’t know that Homeworks had a direct IP connection.  The IP and port method may work with it, but telnet is not the same as a raw connection.  You may need to fiddle with ‘\n’ and/or ‘\r’ in the code.
If that doesn’t work, another idea is to use pyserial, either in the pyhomeworks code or as a shim process.  The shim would basically listen on a socket, open the serial port, and pass everything read or written.
- Mike

On Sep 16, 2023, at 2:59 PM, Mike Morton ***@***.***> wrote:



I have both a regular IP connection to my Homeworks 8 system an a straight serial connection. However it looks like this solution only support ethernet over serial?

Is there a way in the configuration to get it the straight telnet IP address? Something:

192.168.1.23
Port: 23

?

**Comment by mkirsten** [comment]
Hi, I have exactly the same setup. I have a Lutron Series 8 processor with both serial port and ethernet port. I can Telnet into it (it requires l/p) and issue commands as per e.g., https://assets.lutron.com/a/documents/hwi%20rs232%20protocol.pdf

So the below is, after enabling monitoring of keypads and dimmers; me issuing a dim command (10% dim over 3 seconds with 0 seconds delay), dim response, later dim update (to 0%) and a Keypad LED state. I assume this is similar as per the the RS232 spec. 

`FADEDIM,10,3,0,[01:01:04:02:04]\n
DL, [01:01:04:02:04],  10\n
DL, [01:01:04:02:04],   0\n
KLS, [01:06:22], 000010000000000000000000`

Can I help contribute somehow to make this package also be able to connect through Telnet? I've built a number of smart home integrations myself before, but almost exclusively in Java.

**Comment by dubnom** [comment]
Sure!  I suppose there are two different approaches -
1. Support Pyserial
2. Supper telnet
Both would be preferred.
So, have 3 stream-like classes - TCP, telnet, serial and pass one into the homeworks object.  (As long as the format is the same).
I’ll help, but I don’t have either of the new connection methods, so I can’t test.
After that, the HA Homeworks code will need some tweaking.
- Mike
Sent from my iPhone

On Sep 17, 2023, at 3:18 PM, mkirsten ***@***.***> wrote:

﻿

Hi, I have exactly the same setup. I have a Lutron Series 8 processor with both serial port and ethernet port. I can Telnet into it (it requires l/p) and issue commands as per e.g., https://assets.lutron.com/a/documents/hwi%20rs232%20protocol.pdf

So the below is, after enabling monitoring of keypads and dimmers; me issuing a dim command (10% dim over 3 seconds with 0 seconds delay), dim response, later dim update (to 0%) and a Keypad LED state. I assume this is similar as per the the RS232 spec.

FADEDIM,10,3,0,[01:01:04:02:04] DL, [01:01:04:02:04], 10 DL, [01:01:04:02:04], 0 KLS, [01:06:22], 000010000000000000000000

Can I help contribute somehow to make this package also be able to connect through Telnet? I've built a number of smart home integrations myself before, but almost exclusively in Java.

**Comment by mkirsten** [comment]
Hi, great. Maybe I can contribute with the Telnet then as I can easily test it here. What would be the best way for me to get started, do you think? I know Java well (but I ended my programming career about 20 years ago, so...), home automation and network well and no Python. Is there a way where I can easily get it running locally with some simple test?

**Comment by dubnom** [comment]
Before any code is written, did you try to use port 23 and the telnet ip address of the Homeworks device?
- Mike

On Sep 18, 2023, at 2:22 AM, mkirsten ***@***.***> wrote:



Hi, great. Maybe I can contribute with the Telnet then as I can easily test it here. What would be the best way for me to get started, do you think? I know Java well (but I ended my programming career about 20 years ago, so...), home automation and network well and no Python. Is there a way where I can easily get it running locally with some simple test?

**Comment by mkirsten** [comment]
Great point. However, the Telnet session requires login and hence the connection needs to await LOGIN> and then respond back with "<username>,<password>" back before the commands that I've seen in the Pyhon package can be executed. Can I find a simple way to test the Python code locally, I could try just sending the login string before PROMPTOFF (row 81 in pyhomeworks.py)

**Comment by dubnom** [comment]
Give that a try, it would simplify some things.  Basically, we could add a login option to the existing code and that would fix things for most people.
There are a few ways to test locally.  One is to write a small python program that initiates a connection using a local copy of the pyhomeworks library.  This is already done using ’test.py’ in the source directory.
Another is to futz with it inside Home Assistant, which is a bit more cumbersome, but I can go through the steps if you need.

- Mike

On Sep 18, 2023, at 1:51 PM, mkirsten ***@***.***> wrote:



Great point. However, the Telnet session requires login and hence the connection needs to await LOGIN> and then respond back with "," back before the commands that I've seen in the Pyhon package can be executed. Can I find a simple way to test the Python code locally, I could try just sending the login string before PROMPTOFF (row 81 in pyhomeworks.py)

**Comment by mkirsten** [comment]
Sounds good. I’ll try getting started writing a small wrapper around to test locally then outside HA from that directory. If there’s any particular things to consider to get that working, let me know, otherwise I’ll give it a shot!On 18 Sep 2023, at 20:22, Michael Dubno ***@***.***> wrote:﻿
Give that a try, it would simplify some things.  Basically, we could add a login option to the existing code and that would fix things for most people.

There are a few ways to test locally.  One is to write a small python program that initiates a connection using a local copy of the pyhomeworks library.  This is already done using ’test.py’ in the source directory.

Another is to futz with it inside Home Assistant, which is a bit more cumbersome, but I can go through the steps if you need.



- Mike



On Sep 18, 2023, at 1:51 PM, mkirsten ***@***.***> wrote:







Great point. However, the Telnet session requires login and hence the connection needs to await LOGIN> and then respond back with "," back before the commands that I've seen in the Pyhon package can be executed. Can I find a simple way to test the Python code locally, I could try just sending the login string before PROMPTOFF (row 81 in pyhomeworks.py)

**Comment by mkirsten** [comment]
I have something that works for Telnet now, although only the monitoring so far. What would be best - do you want me to amend the current file to support telnet in addition to socket (e.g., when username + password is given), or is it better I implement it in a new class?

**Comment by dubnom** [comment]
I haven’t looked at this project since I wrote it, but it seems that there are a number of forks that appear to be more functional/modern/better than what I originally wrote.  Login seems to be just one of the features.  I’ll put in some time to integrate these forks and fix documentation if the owners make pull requests.  In the interim, check out their forks before we start a new one.
- Mike

On Sep 19, 2023, at 5:03 PM, mkirsten ***@***.***> wrote:



I have something that works for Telnet now, although only the monitoring so far. What would be best - do you want me to amend the current file to support telnet in addition to socket (e.g., when username + password is given), or is it better I implement it in a new class?

**Comment by mikmort** [comment]
> I have something that works for Telnet now, although only the monitoring so far. What would be best - do you want me to amend the current file to support telnet in addition to socket (e.g., when username + password is given), or is it better I implement it in a new class?

mkirsten -- Fantastic that you got the basis going! I'm happy to help as well, whether that be testing or development.

**Comment by dubnom** [comment]
Excellent!  Also, if anyone wants to assume some ownership on this project, please reach out to me.

**Comment by mkirsten** [comment]
Love to help, but I'm too bad a Python to take lead. However, I just hacked up a small test with an integration to HA and will start working on getting something to work now as a plugin. @dubnom - let me know if we should look at something together to integrate Telnet into the existing package, that may be better for the community. O/w I'll create a simple PoC with Telnet as a standalone integration to test things out with.

**Comment by emontnemery** [comment]
Home Assistant Core 2024.8.0 beta, which was released yesterday, has support for telnet login based on the modifications by @yozik04

Please try it and comment on if it works or not.
